src/execution_plan/execution_plan_reduce_to_apply.c

```diff
@@ -0,0 +1,130 @@
+		// parser-generated constant strings.
+		ExecutionPlan_BoundVariables(plan->root, bound_vars);
+		// Prepare the variables for populating the Argument ops we will build.
+		arguments = ExecutionPlan_BuildArgumentModifiesArray(bound_vars);
+	}
+
+	// Initialize an ExecutionPlan that shares this plan's Record mapping.
+	ExecutionPlan *match_branch_plan = ExecutionPlan_NewEmptyExecutionPlan();
+	ExecutionPlan_AppendSubExecutionPlan(plan, match_branch_plan);
+
+	// We have bound variables, build an Argument op that represents them.
+	OpBase *argument = NewArgumentOp(match_branch_plan, arguments);
+	match_branch_plan->root = argument;
+
+	// Build a temporary AST holding a MATCH clause.
+	AST *match_branch_ast = AST_MockMatchPattern(ast, path);
+
+	ExecutionPlan_PopulateExecutionPlan(match_branch_plan, NULL, NULL, NULL);
+
+	AST_MockFree(match_branch_ast);
+	QueryCtx_SetAST(ast); // Reset the AST.
+
+	return match_branch_plan->root;
+}
+
+
+/* Swap operation on operation children. If two valid indices, a and b, are given, this operation
+ * swap the child in index a with the child in index b. */
+static void _OpBaseSwapChildren(OpBase *op, int a, int b) {
+	assert(a >= 0 && b >= 0 && a < op->childCount && b < op->childCount);
+	OpBase *tmp = op->children[a];
+	op->children[a] = op->children[b];
+	op->children[b] = tmp;
+}
+
+static void _CreateBoundBranch(OpBase *op, ExecutionPlan *plan) {
+	rax *bound_vars = NULL;
+	const char **arguments = NULL;
+	if(plan->root) {
+		bound_vars = raxNew();
+		// Rather than cloning the record map, collect the bound variables along with their
+		// parser-generated constant strings.
+		ExecutionPlan_BoundVariables(plan->root, bound_vars);
+		// Prepare the variables for populating the Argument ops we will build.
+		arguments = ExecutionPlan_BuildArgumentModifiesArray(bound_vars);
+	}
+	OpBase *arg = NewArgumentOp(plan, arguments);
+
+	/* In case of Apply operations, we need to append, the new argument branch and
+	 * set it as the first child. */
+	ExecutionPlan_AddOp(op, arg);
+
+	if(op->type & (OPType_APPLY_MULTIPLEXER | OPType_SEMI_APPLY))
+		_OpBaseSwapChildren(op, 0, op->childCount - 1);
+}
+
+/* This method reduces a filter tree into an OpBase. The method perfrom post-order traversal over the
+ * filter tree, and checks if if the current subtree rooted at the visited node contains path filter or not,
+ * and either reduces the root or continue traversal and reduction.
+ * The three possible operations could be:
+ * 1. OpFilter - In the case the subtree originated from the root does not contains any path filters.
+ * 2. OpSemiApply - In case of the current root is an expression which contains a path filter.
+ * 3. ApplyMultiplexer - In case the current root is an operator (OR or AND) and at least one of its children
+ * has been reduced to OpSemiApply or ApplyMultiplexer. */
+static OpBase *_ReduceFilterToOpBase(ExecutionPlan *plan,
```

Comment:
consider renaming to `_ReduceFilterToOp`

Reply:
done